cpp/src/arrow/util/basic_decimal.cc

```diff
@@ -395,6 +395,34 @@ BasicDecimal128& BasicDecimal128::operator*=(const BasicDecimal128& right) {
   return *this;
 }
 
+/// Expands the given big endian array of uint64_t into an array of uint32_t.
+/// The value of input array is expected to be positive. The result_array will
+/// remove leading zeros from the input array.
+/// \param value_array an big endian array to represent the value
+/// \param result_array an array of length N*2 to set with the value
+/// \result the output length of the array
+template <size_t N>
+static int64_t FillInArray(std::array<uint64_t, N>& value_array, uint32_t* result_array) {
+  int64_t next_index = 0;
+  for (size_t i = 0; i < N; i++) {
+    if (value_array[i] == 0) {
+      if (next_index != 0) {
+        result_array[next_index++] = 0;
+        result_array[next_index++] = 0;
+      }
+    } else if (value_array[i] <= std::numeric_limits<uint32_t>::max()) {
```

Comment:
This special case is needed here as we want to remove the leading zero in the result array of uint32_t, so that when the first non-zero uint64_t value is less than the max of uint32_t, we assign the value to the first element of the result array; when the non-zero uint64_t value is not the first value, we reserve the next element of the result array with 0;